---
pdf_document:
  latex_engine: xelatex
author: "Yogita Sharma and Per Ludvik Bratt√•s"
date: "March, 2019"
output:
  html_document: default
title: "Deseq analysis of iN.d5 example"
---
  
```{r, setup, include=FALSE}
knitr::opts_chunk$set(echo = T,results = 'hide', eval=F)
```

## Tutorial on DESeq2
## Analysing iN-d5 RNA-sequencing data

### Download public data from NCBI GEO
##### https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE90068

#### 1. Setup of workspace


```{r, setup workspace, results='hide', message = FALSE }

# clean environment and workspace
rm(list=ls())
gc()

# path to project
setwd("~/Dropbox (MN)/RNA-seq workshop/R analysis/")
path <- getwd()

```


#### 2. Read data
```{r, results='hide', message = FALSE}

# path and filename of raw featureCounts file
fc.file <- "Data/GSE90068_hg38.mRNA.NCBI.fCounts.primary.tab.txt"
fc.file.path <- paste(path,fc.file,sep = "/")

# read data - count table from featureCounts
data <- read.csv(file = fc.file.path,header = T,skip = 1)
head(data)
```


#### 3. Define data

```{r, results='hide', message = FALSE}

# We only want C2 NEG CTL and C2 REST_sh.PBPA data
colnames(data)

library(tidyverse)
countdata <- data %>%
  select(contains("C2.NEG.CTL"),contains("C2.REST_sh.PBPA")) 
head(countdata)
rownames(countdata) <- make.names(data$Geneid,unique = T)
head(countdata)

# take the column names of the rawCount table from featureCounts
cnames <- colnames(countdata)

# remove the substrings that are not needed for identifying the samples, noise from filenames
sampleNames <- gsub(pattern = ".trimmedAligned.out.sorted.bam",replacement = "",x = cnames)
sampleNames <- gsub(pattern = "hg38.C2.",replacement = "",x = sampleNames)
sampleNames

# make colnames
colnames(countdata) <- sampleNames

# get read depth of samples
depth <- colSums(countdata)/1e6

#plot read depth of samples
par(mar=c(10,5,5,5),mfrow=c(1,1),las=2)
barplot(depth)


```

### 4. colData data.frame that describes the data
```{r, results='hide', message = FALSE }
# make one vector describing if the sample has REST inhibition or not
condition <- rep(c("CTL","RESTsh_pBpA"),each=3)

# ColData
colData <- data.frame(condition=condition)

colData
```


### 5. Run DESeq
```{r, results='hide', message = FALSE}
#  make a DESeq object
library(DESeq2)

# create object
dds <- DESeqDataSetFromMatrix(countData = countdata,colData = colData,design = ~condition)

# run DESeq
dds <- DESeq(dds)

```

### 6. Quality contorl

#### 6.1 Peak at data

```{r, results='hide', message = FALSE}

## Get normalized count data
nCounts <- counts(dds,normalized=T)
head(nCounts)

## Calculate average coutns in each condition
aCounts <- sapply( levels(dds$condition), function(lvl) rowMeans( counts(dds,normalized=TRUE)[,dds$condition == lvl] ) )
head(aCounts)

aCounts["SOX2",]
aCounts["ASCL1",]
```

#### 6.2 PCA
```{r, results='hide', message = FALSE}

## Log2 Transform the data
vst <- varianceStabilizingTransformation(dds,blind = T)
avst <- data.frame(assay(vst))
head(avst)

# pca
pca <- prcomp(t(avst))
head(pca)
head(pca$x)
pcax <- data.frame(pca$x)
pcax$condition <- rep(c("CTL","REST"),each=3)

library(ggplot2)
ggplot(data = data.frame(pca$x),mapping = aes(x = PC1,y = PC2,color=condition)) +
  geom_point(size=4) +
  theme_classic() +
  ggtitle(label = "PCA")


```

##### 6.3 Sample to Sample distances 

```{r,  results='hide', message = FALSE}
# compute distances (eucledian) between each sample
sampleDist <- dist(t(avst))
sampleDistMatrix <- as.matrix(sampleDist)
colnames(sampleDistMatrix) <- NULL

# make colors 
colors <- colorRampPalette(brewer.pal(11,"RdBu"))(100)

# plot 
pheatmap(sampleDistMatrix,
         clustring_distance_rows=sampleDist,
         clustring_distance_cols=sampleDist,
         col=colors,
         main = "Sample to Sample distances")

```

##### 6.4 Heatmaps - check variability of data
```{r, results='hide', message = FALSE}
# heatmaps
library(pheatmap)
library(RColorBrewer)

## Visualize most variable genes
avst.sd <- apply(avst,1,sd)
head(avst)
avst <- avst[order(-avst.sd),]
head(avst)

color <- rev(colorRampPalette(brewer.pal(10,"RdBu"))(200))
pheatmap(avst[1:10,],color = color)
pheatmap(avst[1:50,],color = color)
pheatmap(avst[1:250,],color = color,show_rownames = F)
pheatmap(avst[1:1000,],color = color,show_rownames = F)

# visualize a defined set of genes ..

neuronal.genes <- c("NPPC","POU3F2","ASCL1","CLDN5","HMX1","KCNN3","POU3F1","BMP3","PAWR","OPRK1","RAPGEF4","GBA","DRD1","DKK1")
pheatmap(avst[neuronal.genes,],color = color,show_rownames = T)


```

#### 7. Differential expression

```{r, results='hide', message = FALSE}
# get results from diffex test
res <- results(dds)
res

res["POU3F2",]
res["ASCL1",]

# create new variable for plotting
res.df <- data.frame(res)
plotData <- res.df

# log2 transform baseMean levels 
plotData$log2baseMean <- log2(plotData$baseMean+1)
# take -log10 of adjusted p-values (smooth for visualization)
plotData$log10padj <- -log10(plotData$padj)

p <- ggplot(data = plotData,mapping = aes(x = log2baseMean,y = log2FoldChange)) 
p 
p + geom_point()

# change style
p +
  geom_point() +
  theme_classic()

# change size of points
p +
  geom_point(size=.5) +
  theme_classic()


# color and size based on significane 
p +
  geom_point(mapping = aes(size=log10padj,color=log10padj)) +
  theme_classic() +
  scale_size_continuous(range = c(0,3)) 

# alpha - transparency
p +
  geom_point(mapping = aes(size=log10padj,color=log10padj,alpha=log10padj)) +
  theme_classic() +
  scale_size_continuous(range = c(0,5)) 

# coloring gradients 
library(RColorBrewer)
p +
  geom_point(mapping = aes(size=log10padj,color=log10padj,alpha=log10padj)) +
  theme_classic() +
  scale_size_continuous(range = c(0,5)) +
  scale_colour_gradient2(low = "darkblue", mid = "red",
                         high = "darkred", midpoint = 35, space = "Lab",
                         na.value = "grey50", guide = "colourbar", aesthetics = "colour")

# add title

p +
  geom_point(mapping = aes(size=log10padj,color=log10padj,alpha=log10padj)) +
  theme_classic() +
  scale_size_continuous(range = c(0,5)) +
  scale_colour_gradient2(low = "darkblue", mid = "red",
                         high = "darkred", midpoint = 35, space = "Lab",
                         na.value = "grey50", guide = "colourbar", aesthetics = "colour") +
  ggtitle(label = "RESTi + pB.pA vs Control")


## plot based on significant genes
head(plotData)
plotData$signUP <- ifelse(test = !is.na(plotData$padj) & plotData$padj<0.05 & plotData$log2FoldChange>0,yes = 1,no = 0)
plotData$signDOWN <- ifelse(test = !is.na(plotData$padj) & plotData$padj<0.05 & plotData$log2FoldChange<0,yes = 1,no = 0)

# plot all genes, small points
p <- ggplot(data = plotData,mapping = aes(x = log2baseMean,y = log2FoldChange)) +
  geom_point(size=.5,color="black",alpha=.5) +
  theme_classic()
p

## add sign UP and DOWN
p + geom_point(data = plotData[plotData$signUP==1,],
               color= "darkred",size=2,alpha=.5) +
  geom_point(data = plotData[plotData$signDOWN==1,],
             color= "darkblue",size=2,alpha=.5)


## change cutoff
padj <- 0.0001
plotData$signUP <- ifelse(test = !is.na(plotData$padj) & plotData$padj<padj & plotData$log2FoldChange>0,yes = 1,no = 0)
plotData$signDOWN <- ifelse(test = !is.na(plotData$padj) & plotData$padj<padj & plotData$log2FoldChange<0,yes = 1,no = 0)

# plot all genes, small points
p <- ggplot(data = plotData,mapping = aes(x = log2baseMean,y = log2FoldChange)) +
  geom_point(size=.5,color="black",alpha=.5) +
  theme_classic()
p

## add sign UP and DOWN
p + 
  geom_point(data = plotData[plotData$signUP==1,],
             color= "darkred",size=2,alpha=.6) +
  geom_point(data = plotData[plotData$signDOWN==1,],
             color= "darkblue",size=2,alpha=.6) +
  ggtitle(label = "RESTi + pB.pA",subtitle = paste("Sign genes, p-adj<",padj,sep = ""))

```

#### 8. unbiased test for alterations in a gene set 
```{r, results='hide',message=F}
head(plotData)

# make vector with genes of interest
neuronal.genes <- c("NPPC","POU3F2","ASCL1","CLDN5","HMX1","KCNN3","POU3F1","BMP3","PAWR","OPRK1","RAPGEF4","GBA","DRD1","PER1B","DKK1")

# define if each gene is in the given gene set or not
plotData$geneSet <- ifelse(test = rownames(plotData) %in% neuronal.genes,yes = "Neuronal",no = "Other")
head(plotData)
plotData[plotData$geneSet=="Neuronal",]

# plot cumulative fraction of log2FC of the two gene sets
ggplot(plotData,aes(x=log2FoldChange,color=geneSet)) + 
  geom_step(stat="ecdf") +
  theme_classic() +
  ggtitle(label = "Fold changes",subtitle = "Neuronal genes vs all") +
  ylab(label = "Cumulative fraction") 

# plot maPlot of gene sets
ggplot(plotData,aes(x=log2baseMean,y=log2FoldChange)) + 
  geom_point(size=0.5) +
  theme_classic() +
  geom_point(data = plotData[plotData$geneSet=="Neuronal",],color="Red",size=3)

```

#### 9. Write data to file
```{r, results='hide', message = FALSE }

### write sign genes to file
# write the deseq statistics

# all sign genes 
write.table(x = res.df[!is.na(res.df$padj) & res.df$padj<padj,],file = paste("allSignGenes.",padj,".txt",sep = ""))

# print normalized count data
sign.up.names <- rownames(plotData[plotData$signUP==1,])
sign.down.names <- rownames(plotData[plotData$signDOWN==1,])

# write the average normalized read counts of each sign. changed gene 
head(aCounts)

aCounts.up <- aCounts[sign.up.names,]
head(aCounts.up)
aCounts.down <- aCounts[sign.down.names,]
head(aCounts.down)

write.table(x = aCounts.down,file = paste("Count.average_Sign.down.padj.",padj,".txt",sep = ""))
write.table(x = aCounts.up,file = paste("Count.average_Sign.up.padj.",padj,".txt",sep = ""))

```
